onnxruntime/core/providers/webgpu/math/matmul_packed.cc

```diff
@@ -0,0 +1,293 @@
+      << "  let num_tiles = (uniforms.dim_inner - 1) / tileInner + 1;\n"
+      << "  var kStart = 0;\n"
+      << "  var acc: array<vec4<" << data_type << ">, rowPerThread>;\n";
+
+  // Loop over shared dimension.
+  shader.MainFunctionBody()
+      << "  let tileRowB = localRow * " << row_per_thread_b << ";\n"
+      << "  for (var t = 0; t < num_tiles; t = t + 1) {\n";
+
+  // Load one tile of A into local memory.
+  shader.MainFunctionBody()
+      << "    for (var innerRow = 0; innerRow < rowPerThread; innerRow = innerRow + 1) {\n"
+      << "      let inputRow = tileRow + innerRow;\n"
+      << "      let inputCol = tileCol;\n"
+      << "      mm_Asub[inputRow][inputCol] = mm_readA(batch, globalRow + innerRow, kStart / innerElementSize + inputCol, batchIndices);\n"
+      << "    }\n";
+
+  // Load one tile of B into local memory.
+  shader.MainFunctionBody()
+      << "    for (var innerRow = 0; innerRow < " << row_per_thread_b << "; innerRow = innerRow + 1) {\n"
+      << "      let inputRow = tileRowB + innerRow;\n"
+      << "      let inputCol = tileCol;\n"
+      << "      mm_Bsub[inputRow][inputCol] = mm_readB(batch, kStart + inputRow, globalCol, batchIndices);\n"
+      << "    }\n"
+      << "    kStart = kStart + tileInner;\n"
+      << "    workgroupBarrier();\n";
+
+  // Compute acc values for a single thread.
+  shader.MainFunctionBody()
+      << "    for (var k = 0; k < tileInner / innerElementSize; k = k + 1) {\n"
+      << "      let BCached0 = mm_Bsub[k * innerElementSize][tileCol];\n"
+      << "      let BCached1 = mm_Bsub[k * innerElementSize + 1][tileCol];\n"
+      << "      let BCached2 = mm_Bsub[k * innerElementSize + 2][tileCol];\n";
+
+  if (inner_elements_size != 3) {
+    shader.MainFunctionBody() << "      let BCached3 = mm_Bsub[k * innerElementSize + 3][tileCol];\n";
+  }
+
+  shader.MainFunctionBody()
+      << "      for (var i = 0; i < rowPerThread; i = i + 1) {\n"
+      << "        let ACached = mm_Asub[tileRow + i][k];\n"
+      << "        acc[i] = BCached0 * ACached.x + acc[i];\n"
+      << "        acc[i] = BCached1 * ACached.y + acc[i];\n"
+      << "        acc[i] = BCached2 * ACached.z + acc[i];\n"
+      << "        " << (inner_elements_size == 3 ? "" : "acc[i] = BCached3 * ACached.w + acc[i];") << "\n"
+      << "      }\n";
+
+  shader.MainFunctionBody() << "    workgroupBarrier();\n"
+                            << "  }\n";  // main for loop
+
+  // Write the results to the output buffer
+  shader.MainFunctionBody()
+      << "  for (var innerRow = 0; innerRow < rowPerThread; innerRow = innerRow + 1) {\n"
+      << "    mm_write(batch, globalRow + innerRow, globalCol, acc[innerRow]);\n"
+      << "  }\n";
+
+  return Status::OK();
+}
+
+Status MatMulProgram::MakeMatMulPackedSource(ShaderHelper& shader, const ShaderIndicesHelper& batch_dims) const {
+  const int elements_per_thread_x = elements_per_thread_[0];
+  const int elements_per_thread_y = elements_per_thread_[1];
```

Comment:
In order to avoid int vs int64_t type mismatch, use auto. I ran into the type mismatch error on Windows. Also we could use `decltype` to declar new variables as 
`decltype(elements_per_threads_x) tile_inner = 32;`